### PR TITLE
Add $correlation logic.

### LINKED ISSUE #876: Implement correlation RFC
**Describe the problem you are trying to solve**

Implement the Correlation [RFC](https://github.com/tremor-rs/tremor-rfcs/pull/44)

### PR DESCRIPTION
**Pull request by mfelsche**
# Pull request

## Related

* [RFC](https://github.com/tremor-rs/tremor-rfcs/pull/44)
* Related Issues: fixes #876
* Related [docs PR](https://github.com/tremor-rs/tremor-www-docs/pull/100)

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [ ] The performance impact of the change is measured (see below)

### CODE REVIEW

**Comment by mfelsche** [comment]
This one is ready for review/merge. It has coverage pending, which didnt finish properly last night, due to i dont know.

File: src/sink/rest.rs [hunk 1]
```diff
@@ -976,12 +993,13 @@ async fn build_response_events<'response>(
...
     mut error_id: EventId,
     event_id: &EventId,
+    correlation: Option<Value<'static>>,
     status: u16,
     origin_uri: &EventOriginUri,
     e: &Error,
 ) -> Event {
     let mut error_data = Object::with_capacity(2);
-    let mut meta = Object::with_capacity(2);
+    let mut meta = Object::with_capacity(2 + if correlation.is_some() { 1 } else { 0 });
```
Comment (Licenser) [inline src/sink/rest.rs:1002]: we could safely use 3 here. Below 32 it's just a vector allocation.

File: src/sink/rest.rs [hunk 2]
```diff
@@ -1274,6 +1296,7 @@ mod test {
             &mut response_id_gen,
             &event_origin_uri,
             metadata,
+            Some(&Value::Static(StaticNode::I64(-1))),
```
Comment (Licenser) [inline src/sink/rest.rs:1299]: Does a correlation of `-1` have a specific significance here?

Comment (Licenser) [reply src/sink/rest.rs:1299]: It does because this is a test :facepalm: darn compressed diff view ..., sorry.

**Comment by Licenser** [comment]
It would be nice to add correlation to the KV sink as well otherwise, aside of coverage, :+1:

**Comment by mfelsche** [comment]
Damn, i forgot about kv!
Yeah, gonna fix stupid coverage... :(

**Review by Licenser — APPROVED** [review]

**Pull request merged**